orrery/include/orrery/executor.h

```diff
@@ -0,0 +1,67 @@
+// Concord
+//
+// Copyright (c) 2021 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the
+// "License").  You may not use this product except in compliance with the
+// Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the LICENSE
+// file.
+
+#pragma once
+
+#include <thread>
+
+#include "Logger.hpp"
+
+#include "component.h"
+#include "environment.h"
+#include "mailbox.h"
+#include "queue.h"
+
+namespace concord::orrery {
+
+// An executor runs component handler callbacks. It is the actively running thread of control.
+//
+// An executor is always a single thread, although it can proxy work for a thread pool,
+// external process, or even remote compute cluster as necessary.
+//
+// An executor owns a mailbox and its underlying queue. It is driven forward by pulling messages off
+// the queue after they are placed in the mailbox.
+class Executor {
+ public:
+  Executor(const std::string& name) : name_(name), queue_(std::make_shared<detail::Queue>()), mailbox_(name, queue_) {}
+
+  Mailbox& mailbox() { return mailbox_; }
+
+  void init(const Environment& environment) {
+    // The value doesn't matter. `enumSize` is a generated function that ignores the parameter.
+    ComponentId dummy = ComponentId::broadcast;
+    components_.resize(enumSize(dummy));
+    environment_ = environment;
+  }
```

Comment:
Since the queue is shared anyway, I guess who "owns" it doesn't really matter. Maybe it can be a member of `Mailbox` and then we can construct a `Mailbox` separately, passing it  and `Environment` to `Executor`, in order to avoid `init()`. Just a thought. Let me know if it makes sense.

```
 auto mailbox1 = Mailbox("replica_executor");
 auto mailbox2 = Mailbox("state_transfer_executor");

 auto env = Environment();
 env.add(ComponentId::replica, mailbox1);
 env.add(ComponentId::state_transfer, exec2);

 auto exec1 = Executor(env , mailbox1);
 auto exec2 = Executor(env , mailbox2);

```

Reply:
I guess the problem now is that one can pass the same mailbox to two or more executors. However, the user can also mess up the call to Environment::add() by passing the wrong mailbox.

Reply:
I'm open either way. I don't remember the exact reason I did it this way.